tests/ssl-opt.sh

```diff
@@ -669,9 +669,28 @@ requires_openssl_next() {
     fi
 }
 
+# skip next test if openssl version is lower than 3.0
+requires_openssl_3_x() {
+    requires_openssl_next
+    if [ "$OPENSSL_NEXT_AVAILABLE" = "NO" ]; then
+        OPENSSL_3_X_AVAILABLE="NO"
+    fi
+    if [ -z "${OPENSSL_3_X_AVAILABLE:-}" ]; then
+        if $OPENSSL_NEXT version 2>&1 | grep "OpenSSL 3." >/dev/null
+        then
+            OPENSSL_3_X_AVAILABLE="YES"
+        else
+            OPENSSL_3_X_AVAILABLE="NO"
+        fi
+    fi
+    if [ "$OPENSSL_3_X_AVAILABLE" = "NO" ]; then
+        SKIP_NEXT="YES"
+    fi
+}
+
 # skip next test if tls1_3 is not available
 requires_openssl_tls1_3() {
-    requires_openssl_next
+    requires_openssl_3_x
```

Comment:
What the function does is no longer consistent with the name of the function of its description. What is it we're trying to achieve here?

Reply:
When openssl version is less than 3 it can only handle EC groups (without FFDH groups). So `MBEDTLS_ECDH_C` is required to do handshake I think.

Reply:
Ok, but I don't think modifying the existing `requires_openssl_tls1_3` function is the right way to handle that. This function is already widely used, with a well-defined semantic, and we shouldn't change it or we'll end up skipping tests that we previously not skipped.

Instead, I think what we need is one or more new functions like (for example, I'm not sure this is the right one, but just to give the idea) `requires_openssl_tls1_3_with_compatible_ephemeral` that would
1. call `requires_openssl_next`
2. check if we have support for ECDH (`PSA_WANT_ALG_ECDH`)
3. if yes, do nothing, else call `requires_openssl_tls1_3_with_ffdh`
We would use this function to guard test cases that are trying to do an ephemeral handshake that will use ECDH or FFDH depending on what's available.

The function `requires_openssl_tls1_3_with_ffdh` would just call `requires_openssl_3_x` and we would use it on test cases that will always use FFDH because that's mandated by the parameters used (like, the only allowed groups are FFDH). (We could use `requires_openssl_3_x` directly in those cases, but having the `requires_openssl_tls1_3_with_ffdh` wrapper makes the logic more explicit.)

Reply:
So new functions should look like:
```
# skip next test if openssl does not support ffdh keys
requires_openssl_tls1_3_with_ffdh() {
    requires_openssl_3_x
}

# skip next test if it cannot handle ephemeral key exchange
requires_openssl_tls1_3_with_compatible_ephemeral() {
    requires_openssl_next

    if !(is_config_enabled "PSA_WANT_ALG_ECDH"); then
        requires_openssl_tls1_3_with_ffdh
    fi
}

# skip next test if tls1_3 is not available
requires_openssl_tls1_3() {
    requires_openssl_next
    if [ "$OPENSSL_NEXT_AVAILABLE" = "NO" ]; then
        OPENSSL_TLS1_3_AVAILABLE="NO"
    fi
    if [ -z "${OPENSSL_TLS1_3_AVAILABLE:-}" ]; then
        if $OPENSSL_NEXT s_client -help 2>&1 | grep tls1_3 >/dev/null
        then
            OPENSSL_TLS1_3_AVAILABLE="YES"
        else
            OPENSSL_TLS1_3_AVAILABLE="NO"
        fi
    fi
    if [ "$OPENSSL_TLS1_3_AVAILABLE" = "NO" ]; then
        SKIP_NEXT="YES"
    fi
}
```

And now the following adaptations are required:
1. Generated ffdh tests should use `requires_openssl_tls1_3_with_ffdh` instead `requires_openssl_3_x`
2. In `ssl-opt.sh` all test cases that have `MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED` config enable should have also `requires_openssl_tls1_3_with_compatible_ephemeral`

Reply:
That looks correct to me! Hoping I didn't miss anything important...